### 문제1: 114쪽 문제 3.5

웹사이트 www.stat.ufl.edu/~aa/cat/data 로부터 <표 3.2>에 있는 참게 자료를 다운받을 수 있다. 만일 참게가 적어도 한 마리의 부수체를 가지면 y=1이라 하고 그렇지 않으면 y=0이라 하자.

In [59]:
# crab은 연번
# sat는 부수체의 수
# weight는 무게
# width는 등딱지너비
# color는 색깔(1=약간밝은색, 2=중간색, 3=약간어두운색, 4=어두운색)
# spine은 등뼈의 상태(1=둘다양호함, 2=1개가 휘거나 좋지않음, 3=둘다 휘거나 좋지않음)
# y값도 이미 입력되어 있다.

crab_data = read.table("Crabs.dat.txt", header=TRUE)
head(crab_data)

crab sat y weight width color spine
1 1    8   1 3.05   28.3  2     3    
2 2    0   0 1.55   22.5  3     3    
3 3    9   1 2.30   26.0  1     1    
4 4    0   0 2.10   24.8  3     3    
5 5    4   1 2.60   26.0  3     3    
6 6    0   0 2.10   23.8  2     3

In [60]:
summary(crab_data)

      crab          sat               y              weight          width     
 Min.   :  1   Min.   : 0.000   Min.   :0.0000   Min.   :1.200   Min.   :21.0  
 1st Qu.: 44   1st Qu.: 0.000   1st Qu.:0.0000   1st Qu.:2.000   1st Qu.:24.9  
 Median : 87   Median : 2.000   Median :1.0000   Median :2.350   Median :26.1  
 Mean   : 87   Mean   : 2.919   Mean   :0.6416   Mean   :2.437   Mean   :26.3  
 3rd Qu.:130   3rd Qu.: 5.000   3rd Qu.:1.0000   3rd Qu.:2.850   3rd Qu.:27.7  
 Max.   :173   Max.   :15.000   Max.   :1.0000   Max.   :5.200   Max.   :33.5  
     color           spine      
 Min.   :1.000   Min.   :1.000  
 1st Qu.:2.000   1st Qu.:2.000  
 Median :2.000   Median :3.000  
 Mean   :2.439   Mean   :2.486  
 3rd Qu.:3.000   3rd Qu.:3.000  
 Max.   :4.000   Max.   :3.000  

#### a. 무게를 예측변수로 이용해서 선형확률모형을 P(Y=1)에 적합하라. 만약 소프트웨어가 이항분포에 대해서 항등연결함수를 사용할 수 없거나 수렴하지 않으면 Y를 정규분포를 따르는 것처럼 간주하여 보통의 최소제곱법을 이용하여 모수를 추정하고 추정값들을 해석하라. 가장 큰 무게값 5.20 kg에 대하여 P(Y=1)을 예측하고 결론을 내려라.

In [61]:
model1 = lm(y~weight, data=crab_data)
summary(model1)


Call:
lm(formula = y ~ weight, data = crab_data)

Residuals:
    Min      1Q  Median      3Q     Max 
-0.8878 -0.4683  0.1606  0.3704  0.6689 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -0.14487    0.14715  -0.984    0.326    
weight       0.32270    0.05876   5.492 1.42e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.4447 on 171 degrees of freedom
Multiple R-squared:  0.1499,	Adjusted R-squared:  0.1449 
F-statistic: 30.16 on 1 and 171 DF,  p-value: 1.421e-07


lm 함수를 이용하면 무게를 예측변수로 이용해서 선형확률모형을 적합할 수 있다.
적합된 회귀식은 다음과 같다.
<img src="https://latex.codecogs.com/svg.image?\hat{P}(Y=1) = -0.145+0.323\times weight">

적합된 회귀식에 가장 큰 무게값 5.20을 대입하면 
<img src="https://latex.codecogs.com/svg.image?\hat{P}(Y=1) = -0.145+0.323\times 5.20=1.535"> 이다.

예측확률값이 1보다 크다. 즉, 확률이 0~1에 수렴하지 않기 때문에 Y가 정규분포를 따르는 것처럼 간주하여 보통의 최소제곱법을 이용하여 모수를 추정한다.

In [62]:
max_input = data.frame(weight=5.20)

ci = predict(model1, newdata=max_input, interval = "confidence")
ci

fit      lwr      upr     
1 1.533186 1.205838 1.860534

추정된 모수값은 1.533이고, 신뢰구간은 (1.206, 1.861)이다.
추정된 모수값이 1보다 클 확률이 매우 높다고 볼 수 있다.
따라서 "무게가 5.20kg인 경우 거의 확실히 부수체가 존재한다"고 할 수 있다.

#### b. Y를 이항변수로 간주하고 로지스틱 모형을 적합하라. 무게값 5.20 kg에 대하여 P(Y=1) =0.9968 이 됨을 보여라.

In [63]:
model2 = glm(y~weight, data=crab_data, family=binomial)
summary(model2)


Call:
glm(formula = y ~ weight, family = binomial, data = crab_data)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  -3.6947     0.8802  -4.198 2.70e-05 ***
weight        1.8151     0.3767   4.819 1.45e-06 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 225.76  on 172  degrees of freedom
Residual deviance: 195.74  on 171  degrees of freedom
AIC: 199.74

Number of Fisher Scoring iterations: 4


In [64]:
pi_x = predict(model2, newdata=max_input, type="response")
pi_x

1 
0.9968084

### 문제2: 114쪽 문제 3.6

2016년도 일반사회조사에서 18세에서 27세까지를 대상으로 정치성향(1=매우 보수적, 2=보수적, 3=약간 보수적, 4=중간, 5=약간 진보적, 6=진보적, 7=매우 진보적)과 정당(민주당, 공화당)에 대하여 교차분할표를 구한 것이다.

||1|2|3|4|5|6|7|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Democrat|5|18|19|25|7|7|2|
|Republican|1|3|1|11|10|11|1|

지지정당이 민주당일 확률에 미치는 정치성향의 효과를 보기 위해 R을 이용하여 보형을 적합한 결과가 다음과 같다.

In [65]:
y <- c(5,18,19,25,7,7,2); n <- c(6,21,20,36,17,18,3)
x <- c(1,2,3,4,5,6,7)

model3 <- glm(y/n ~ x, family=binomial(link=logit), weights=n)
summary(model3)


Call:
glm(formula = y/n ~ x, family = binomial(link = logit), weights = n)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)   3.1870     0.7002   4.552 5.33e-06 ***
x            -0.5901     0.1564  -3.772 0.000162 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 24.7983  on 6  degrees of freedom
Residual deviance:  7.7894  on 5  degrees of freedom
AIC: 30.516

Number of Fisher Scoring iterations: 4


In [66]:
confint(model3)

Waiting for profiling to be done...



2.5 %     97.5 %    
(Intercept)  1.901798  4.6648412
x           -0.915875 -0.2983235

#### a. 예측식을 제시하고 정치성향 효과의 방향에 대해서 해석하라.

주어진 코드를 보면 y는 민주당 지지자 수이고, n은 전체 조사대상자 수이다.
즉, π(x) = P(Y=1)은 민주당을 지지할 확률을 의미한다.

summary 함수 결과에 따른 예측식은 다음과 같다.

<img src="https://latex.codecogs.com/svg.image?logit[\pi(x)]=log\left[\frac{\pi(x)}{1-\pi(x)} \right]=3.187-0.590 x"><br>

<img src="https://latex.codecogs.com/svg.image?e^{\hat\beta}=e^{-0.590}=0.554">

민주당 지지에 대한 오즈는 정치성향이 1증가함에 따라 약 0.45배 감소한다.
즉, 정치성향이 진보적일수록 민주당 지지 오즈가 낮아진다.

#### b. 정치성향 효과에 대한 95% 왈드 신뢰구간을 구하고 위에 주어진 프로파일 가능도 신뢰구간과 비교하라.

In [67]:
# 신뢰구간을 구하기 위한 데이터 입력
beta = -0.5901
std_error = 0.1564
z = qnorm(0.975)

# 신뢰구간 계산
ci_hi = beta + z * std_error
ci_low = beta - z * std_error

cat(ci_low, ci_hi)

-0.8966384 -0.2835616

정치성향 효과에 대한 왈드 신뢰구간은 (-0.897, -0.284)이고, 프로파일 가능도 신뢰구간은 (-0.916, -0.298) 이다.

#### c. 정치성향 효과에 대한 왈드 검정을 수행하라. 검정통계량값과 P-값을 제시하고 결과를 해석하라.

In [68]:
wald_value = (beta/std_error)^2
p_value = 1-pchisq(wald_value, df=1)

cat("wald_value:", wald_value, "\n")
cat("p-value:", p_value)

wald_value: 14.23566 
p-value: 0.0001612847

왈드 검정의 검정통계량 값은 14.236이며, 이 값은 정치성향이 민주당 지지에 미치는 영향이 얼마나 큰지 나타낸다. 값이 클수록 영향력이 크다고 해석할 수 있다.

P-값은 0.000161로 매우 작은 값이며, 이 값이 작으면 정치성향이 민주당 지지에 영향을 미치지 않는다는 귀무가설을 기각할 수 있는 근거가 된다.

즉, 정치성향 효과에 대한 왈드 검정에 따르면 정치성향은 민주당 지지에 유의미한 영향을 주는 것으로 판단된다.

#### d. 가능도비 검정을 수행하고 검정통계량값과 P-값을 제시하고 결과를 해석하라.

In [69]:
# 단순 모델(β=0 일 때 로그가능도 함수)
model4 = glm(y/n ~ 1, family=binomial(link=logit), weights=n)

# 가능도 계산
L_1 = logLik(model3)
L_0 = logLik(model4)

# 가능도비 검정 통계량
l_value = -2 * (L_0 - L_1)

# p값 계산
p_value = 1-pchisq(l_value, df=1)

# 결과
cat("l_value:", l_value, "\n")
cat("p_value:", p_value)

l_value: 17.00891 
p_value: 3.720474e-05

가능도비 검정의 검정통계량 값은 17.009이며, 이 값이 클수록 β에 대해 아무 제한도 없는 로그가능도 함수의 최대값이 β=0인 제약이 있는 로그가능도 함수의 최대값보다 크다. 즉, 이 값이 클수록 β에 대해 아무 제한도 없는 로그가능도 함수가 데이터에 잘 맞는다는 것을 의미한다.

P-값은 매우 작은 값이며, 이는 정치성향이 민주당 지지에 유의미한 영향을 미친다고 해석할 수 있다.

#### e. R 출력 결과에 있는 Fisher 스코어 알고리즘의 반복수에 대해서 설명하라.

Number of Fisher Scoring iterations는 4인데, 이 값은 일반화선형모델에서 파라미터를 추정하는 과정에서 로그가능도의 기울기를 이용하여 다음 추정치를 업데이트 하는데 이 과정이 몇번이나 반복되었는지를 나타낸다.

모델의 최적화 과정에서 발생하는 반복 수가 적을수록 더 빠르고 효율적인 추정이 이루어졌다고 볼 수 있다.

### 문제3: 117쪽 문제 3.11

컴퓨터 칩의 실리콘 기판을 제조하는 데 사용되는 두 공정 과정의 결합율을 분석하기 위해 실험을 하였다. 10개의 기판에 대해 처리 A를 적용하였을 때에 기판에서 관측된 결함 수는 각각 8, 7, 6, 6, 3, 4, 7, 2, 3, 4 이었다. 10개의 기판에 대하여 처리 B를 적용하였을 때에는 결함 수가 9, 9, 8, 14, 8, 13, 11, 5, 7, 6 이었다. 결함 수를 각각 평균 <img src="https://latex.codecogs.com/svg.image?\mu_A">와 <img src="https://latex.codecogs.com/svg.image?\mu_B">를 갖는 서로 독립인 포아송 변량이라고 간주하자. 모형 <img src="https://latex.codecogs.com/svg.image?log\mu=\alpha+\beta x">을 고려하자. 여기서 처리 A는 x=0, 처리 B는 x=1로 나타낸다. 그러면 <img src="https://latex.codecogs.com/svg.image?\beta=log\mu_B-log\mu_A=log(\mu_B/\mu_A)">이고, <img src="https://latex.codecogs.com/svg.image?e^\beta=\mu_B/\mu_A">가 된다. 이 모형을 적합하고 예측식을 제시하고 <img src="https://latex.codecogs.com/svg.image?\hat\beta">값을 해석하라.

In [70]:
# 결함 수 데이터
data_A <- c(8, 7, 6, 6, 3, 4, 7, 2, 3, 4)
data_B <- c(9, 9, 8, 14, 8, 13, 11, 5, 7, 6)

# 데이터 프레임 생성
data_AB <- data.frame(
  defects = c(data_A, data_B),
  process = factor(c(rep(0, 10), rep(1, 10)))
)

# 포아송 회귀 모델 적합
model5 <- glm(defects ~ process, data = data_AB, family = poisson(link="log"))

# 모델 요약
summary(model5)


Call:
glm(formula = defects ~ process, family = poisson(link = "log"), 
    data = data_AB)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)   1.6094     0.1414  11.380  < 2e-16 ***
process1      0.5878     0.1764   3.332 0.000861 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 27.857  on 19  degrees of freedom
Residual deviance: 16.268  on 18  degrees of freedom
AIC: 94.349

Number of Fisher Scoring iterations: 4


이 모형을 적합한 예측식은 다음과 같다.

<img src="https://latex.codecogs.com/svg.image?\hat\mu=e^{\hat\alpha+\hat\beta x}=e^{1.609+0.588 x}"><br>

<img src="https://latex.codecogs.com/svg.image?\hat\beta"> 값이 0.588인 것은 process 값이 1 커질 때 결함이 발생할 오즈가 <img src="https://latex.codecogs.com/svg.image?e^{\hat\beta}=e^{0.588}=1.8">배 증가한다.

즉, 처리 B를 적용했을 때 결함이 발생할 오즈가 80% 증가한다는 의미이다.

### 문제4: 117쪽 문제 3.12

연습문제 3.11을 참조하라.

#### a. <img src="https://latex.codecogs.com/svg.image?H_0:\mu_A=\mu_B">에 대한 검정을 <img src="https://latex.codecogs.com/svg.image?H_0:\beta=0">의 왈드 검정이나 가능도비 검정을 통하여 실시하라.

In [71]:
# 왈드 검정
beta = summary(model5)$coefficients[2,1]
std_error = summary(model5)$coefficients[2,2]

wald_value = (beta/std_error)^2
p_value = 1-pchisq(wald_value, df=1)

cat("wald_value:", wald_value, "\n")
cat("p-value:", p_value)

wald_value: 11.10517 
p-value: 0.0008608747

왈드 검정의 검정통계량 값은 11.105이며, 이 값은 처리방법이 결함에 미치는 영향이 얼마나 큰지 나타낸다. 값이 클수록 영향력이 크다고 해석할 수 있다.

P-값은 0.000861로 매우 작은 값이며, 이 값이 작으면 처리방법이 결함에 영향을 미치지 않는다는 귀무가설을 기각할 수 있는 근거가 된다.

즉, 처리방법에 대한 왈드 검정에 따르면 처리방법은 결함에 유의미한 영향을 주는 것으로 판단된다.

In [72]:
# 단순 모델(β=0 일 때 로그가능도 함수)
model6 <- glm(defects ~ 1, data = data_AB, family = poisson(link="log"))

# 가능도 계산
L_1 = logLik(model5)
L_0 = logLik(model6)

# 가능도비 검정 통계량
l_value = -2 * (L_0 - L_1)

# p값 계산
p_value = 1-pchisq(l_value, df=1)

# 결과
cat("l_value:", l_value, "\n")
cat("p_value:", p_value)

l_value: 11.58937 
p_value: 0.0006632976

가능도비 검정의 검정통계량 값은 11.589이며, 이 값이 클수록 β에 대해 아무 제한도 없는 로그가능도 함수가 데이터에 잘 맞는다는 것을 의미한다.

P-값은 매우 작은 값이며, 이는 처리방법이 결함에 유의미한 영향을 미친다고 해석할 수 있다.

#### b. <img src="https://latex.codecogs.com/svg.image?\mu_B/\mu_A">에 대한 95% 신뢰구간을 구하라.(힌트: 우선 <img src="https://latex.codecogs.com/svg.image?\beta=log\mu_B-log\mu_A=\log(\mu_B/\mu_A)">에 대한 신뢰구간을 구하라.)

In [73]:
# 신뢰구간을 구하기 위한 데이터 입력
beta = summary(model5)$coefficients[2,1]
std_error = summary(model5)$coefficients[2,2]
z = qnorm(0.975)

# 신뢰구간 계산
ci_hi = beta + z * std_error
ci_low = beta - z * std_error

cat(ci_low, ci_hi)

0.242082 0.9334913

<img src="https://latex.codecogs.com/svg.image?\beta=log(\mu_B/\mu_A)">이다.

우선 β의 신뢰구간을 구하면 (0.242, 0.933)이다.

<img src="https://latex.codecogs.com/svg.image?\beta=log(\mu_B/\mu_A)">이므로 이 값에 지수제곱을 취하면 <img src="https://latex.codecogs.com/svg.image?\mu_B/\mu_A">의 신뢰구간을 구할 수 있다.

In [74]:
ci2_hi = exp(ci_hi)
ci2_low = exp(ci_low)

cat(ci2_low, ci2_hi)

1.273899 2.543373

<img src="https://latex.codecogs.com/svg.image?\mu_B/\mu_A">의 신뢰구간은 (1.274, 2.543) 이다.